<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>prepare_tcr_seq_data • igsc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="prepare_tcr_seq_data">
<meta property="og:description" content="igsc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">igsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9700</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/prepare_tcr_seq_data.html">prepare_tcr_seq_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="prepare_tcr_seq_data_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>prepare_tcr_seq_data</h1>
            
      
      
      <div class="hidden name"><code>prepare_tcr_seq_data.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"randomNames"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"parallel"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"pbapply"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"ggbeeswarm"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">igsc</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></code></pre></div>
<p>Prepare a folder with sub-folder of your samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># vdj_path is the directory on your disk and may contain sub-folders of different samples</span>
<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">vdj_path</span><span class="op">)</span>
<span class="co">#&gt; [1] "Pat3_rep1_blood" "Pat4_rep1_blood"</span></code></pre></div>
<p>Each sub-folder must contain the 6 files from the cr output which are required to prepare a data frame as in the procedure below. Some modifications may be possible putting the functionality of downstream functions at risk though.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">vdj_path</span>, <span class="st">"Pat3_rep1_blood"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "clonotypes.csv"                 </span>
<span class="co">#&gt; [2] "concat_ref.fasta"               </span>
<span class="co">#&gt; [3] "consensus_annotations.csv"      </span>
<span class="co">#&gt; [4] "consensus.fasta"                </span>
<span class="co">#&gt; [5] "filtered_contig_annotations.csv"</span>
<span class="co">#&gt; [6] "filtered_contig.fasta"</span></code></pre></div>
<div id="read-raw-files-from-cellranger-cr-output" class="section level2">
<h2 class="hasAnchor">
<a href="#read-raw-files-from-cellranger-cr-output" class="anchor"></a>Read raw files from cellranger (cr) output</h2>
<p>Read the files of cellrangers VDJ alignment (clonotypes) arranged as described above into one long data frame:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cl_long</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/read_cellranger_outs.html">read_cellranger_outs</a></span><span class="op">(</span><span class="va">vdj_path</span><span class="op">)</span></code></pre></div>
</div>
<div id="attach-the-best-matching-allele-from-imgt" class="section level2">
<h2 class="hasAnchor">
<a href="#attach-the-best-matching-allele-from-imgt" class="anchor"></a>Attach the best matching allele from IMGT</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># optional:</span>
<span class="co"># imgt_ref &lt;- igsc::imgt_tcr_segment_prep()</span>
<span class="co"># see ?igsc::imgt_tcr_segment_prep()</span>
<span class="va">imgt_ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"IMGT_ref/human/hs.rds"</span>, package <span class="op">=</span> <span class="st">"igsc"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># for mouse: imgt_ref &lt;- readRDS(system.file("extdata", "IMGT_ref/mouse/mm.rds", package = "igsc"))</span>

<span class="va">ata</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/attach_imgt_alleles.html">attach_imgt_alleles</a></span><span class="op">(</span>seq <span class="op">=</span> <span class="va">cl_long</span>, imgt_ref <span class="op">=</span> <span class="va">imgt_ref</span>, pick.by <span class="op">=</span> <span class="st">"random"</span>, lapply_fun <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>, mc.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="co"># better set pick.by = "alignment" if you are serious about the results</span>
<span class="va">cl_long</span> <span class="op">&lt;-</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">cl_long</span>, <span class="va">ata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"consensus_seq_cr"</span> <span class="op">=</span> <span class="st">"seq"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyr/man/separate.html">separate</a></span><span class="op">(</span><span class="va">VJ_IMGT</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"V_imgt"</span>, <span class="st">"J_imgt"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"___"</span><span class="op">)</span></code></pre></div>
</div>
<div id="make-a-wide-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#make-a-wide-data-frame" class="anchor"></a>Make a wide data frame</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">collapse.fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">x</span>,collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">cl_wide</span> <span class="op">&lt;-</span>
  <span class="va">cl_long</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/select.html">select</a></span><span class="op">(</span><span class="va">chain</span>, <span class="va">barcode</span>, <span class="va">clonotype_id_cr</span>, <span class="va">sample</span>, <span class="va">CDR3_nt_cr</span>, <span class="va">CDR3_aa_cr</span>, <span class="va">V_cr</span>, <span class="va">J_cr</span>, <span class="va">C_cr</span>, <span class="va">consensus_seq_cr</span>, <span class="va">ref_seq_cr</span>, <span class="va">V_imgt</span>, <span class="va">J_imgt</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyr/man/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">chain</span>, values_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">CDR3_nt_cr</span>, <span class="va">CDR3_aa_cr</span>, <span class="va">V_cr</span>, <span class="va">J_cr</span>, <span class="va">C_cr</span>, <span class="va">consensus_seq_cr</span>, <span class="va">ref_seq_cr</span>, <span class="va">V_imgt</span>, <span class="va">J_imgt</span><span class="op">)</span>, values_fn <span class="op">=</span> <span class="va">collapse.fun</span><span class="op">)</span>
<span class="va">cl_wide</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">)</span></code></pre></div>
</div>
<div id="add-clonotype-names" class="section level2">
<h2 class="hasAnchor">
<a href="#add-clonotype-names" class="anchor"></a>Add clonotype names</h2>
<p>With the randomNames packages you may give your clonotypes a catchy name, as an alternative to e.g. “clonotype1001”. The sampling from randomNames does not work perfect though, so we use a custom function to avoid duplicate names.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "Pat3_rep1_blood" "Pat4_rep1_blood"</span>
<span class="va">cl_wide</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyr/man/separate.html">separate</a></span><span class="op">(</span><span class="va">cl_wide</span>, <span class="va">sample</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"patient"</span>, <span class="st">"replicate"</span>, <span class="st">"body_fluid"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span>, remove <span class="op">=</span> <span class="cn">F</span><span class="op">)</span>
<span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/distinct.html">distinct</a></span><span class="op">(</span><span class="va">cl_wide</span>, <span class="va">CDR3_nt_cr_TRA</span>, <span class="va">CDR3_nt_cr_TRB</span>, <span class="va">patient</span><span class="op">)</span>
<span class="va">temp</span><span class="op">$</span><span class="va">cl_name</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/pick_randomNames.html">pick_randomNames</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span>,
                                       gender <span class="op">=</span> <span class="fl">2</span>,
                                       which.names <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span>
<span class="co">## only temporary to allow comparison below</span>
<span class="va">temp</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">temp</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"___"</span>
<span class="va">temp</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">temp</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"___"</span>
<span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"___"</span>
<span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"___"</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">cl_wide</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">patient</span> <span class="op">==</span> <span class="va">temp</span><span class="op">[</span><span class="va">i</span>,<span class="st">"patient"</span><span class="op">]</span><span class="op">)</span>,
                                 <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span> <span class="op">==</span> <span class="va">temp</span><span class="op">[</span><span class="va">i</span>,<span class="st">"CDR3_nt_cr_TRA"</span><span class="op">]</span><span class="op">)</span>,
                                 <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span> <span class="op">==</span> <span class="va">temp</span><span class="op">[</span><span class="va">i</span>,<span class="st">"CDR3_nt_cr_TRB"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="st">"cl_name"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">temp</span><span class="op">[</span><span class="va">i</span>,<span class="st">"cl_name"</span><span class="op">]</span>
<span class="op">}</span>

<span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRA</span> <span class="op">==</span> <span class="st">"___"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span>
<span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">cl_wide</span><span class="op">$</span><span class="va">CDR3_nt_cr_TRB</span> <span class="op">==</span> <span class="st">"___"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></code></pre></div>
</div>
<div id="paste-clonotype-names-to-cl_long" class="section level2">
<h2 class="hasAnchor">
<a href="#paste-clonotype-names-to-cl_long" class="anchor"></a>Paste clonotype names to cl_long</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cl_long</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">cl_long</span>, <span class="va">cl_wide</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/select.html">select</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">body_fluid</span>, <span class="va">replicate</span>, <span class="va">barcode</span>, <span class="va">sample</span>, <span class="va">cl_name</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="st">"barcode"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Save the processed clonotype data. The wide format may be used to create meta-data columns in the Seurat Object. That way the phenotype data are complemented with clonotype information which are then readily available e.g. as color code on a dimension reduction plot or so.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>cl_wide <span class="op">=</span> <span class="va">cl_wide</span>, cl_long <span class="op">=</span> <span class="va">cl_long</span><span class="op">)</span>
<span class="co">#saveRDS(cl, "clonotype_data.rds")</span></code></pre></div>
</div>
<div id="align-tcr-sequencing-data-with-imgt-references-and-obtain-the-vdj-range" class="section level2">
<h2 class="hasAnchor">
<a href="#align-tcr-sequencing-data-with-imgt-references-and-obtain-the-vdj-range" class="anchor"></a>Align TCR sequencing data with IMGT references and obtain the V(D)J-range</h2>
<p>Now everything is set up to systematically align IMGT TCR reference gene segments to our sequencing data of TCRs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># find an expanded clonotype_name</span>
<span class="va">top_expanded_clonotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">cl</span><span class="op">[[</span><span class="st">"cl_wide"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cl_name</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">top_expanded_clonotypes</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Ebony   Armando    O'Shea    Kaying   Shomari Josephine </span>
<span class="co">#&gt;        33        20        17        16        16        15</span>
<span class="co"># use the TCR that has been detected in most single cell transcriptomes</span>
<span class="co"># check the documentation of align_imgt_ref_to_TCR_seq to understand the inputs</span>
<span class="va">align_list_TRA</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/align_imgt_ref_to_TCR_seq.html">align_imgt_ref_to_TCR_seq</a></span><span class="op">(</span>chain <span class="op">=</span> <span class="st">"TRA"</span>, TCR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">top_expanded_clonotypes</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"cl_name"</span><span class="op">)</span>, cl_wide <span class="op">=</span> <span class="va">cl</span><span class="op">[[</span><span class="st">"cl_wide"</span><span class="op">]</span><span class="op">]</span>, cl_long <span class="op">=</span> <span class="va">cl</span><span class="op">[[</span><span class="st">"cl_long"</span><span class="op">]</span><span class="op">]</span>, imgt_ref <span class="op">=</span> <span class="va">imgt_ref</span><span class="op">)</span>
<span class="va">align_list_TRB</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/align_imgt_ref_to_TCR_seq.html">align_imgt_ref_to_TCR_seq</a></span><span class="op">(</span>chain <span class="op">=</span> <span class="st">"TRB"</span>, TCR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">top_expanded_clonotypes</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"cl_name"</span><span class="op">)</span>, cl_wide <span class="op">=</span> <span class="va">cl</span><span class="op">[[</span><span class="st">"cl_wide"</span><span class="op">]</span><span class="op">]</span>, cl_long <span class="op">=</span> <span class="va">cl</span><span class="op">[[</span><span class="st">"cl_long"</span><span class="op">]</span><span class="op">]</span>, imgt_ref <span class="op">=</span> <span class="va">imgt_ref</span><span class="op">)</span></code></pre></div>
<p>The output of align_imgt_ref_to_TCR_seq contains several plots and sequence information.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">align_list_TRA</span><span class="op">)</span>
<span class="co">#&gt; [1] "consensus.alingments" "consensus.seqs.uncut"</span>
<span class="co">#&gt; [3] "VJC.alignments"       "VJ.alignments"       </span>
<span class="co">#&gt; [5] "V.to.J.TCR.seqs"</span>
<span class="co"># this plot shows how many different unique sequences existed in cl_long for this TCR; source column is consensus_seq_cr</span>
<span class="co"># in different cells (single cell transcriptomes) the consensus_seq_cr may differ slightly but still the TCR is grouped into one clonotype</span>
<span class="co"># sometimes there are irrelevant sequence differences at the beginning or the end, which do not affect the V(D)J-region</span>
<span class="co"># if there are at least 2 different ones the consensus is used for further processing</span>
<span class="va">align_list_TRA</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="prepare_tcr_seq_data_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># the consensus sequence is available at the second index</span>
<span class="va">align_list_TRA</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; [1] "TGGGGTCTATTGGCTGAGTAGACAATGTGAGAAACCACTATGATTGTTAGAGGAAAGAAAGGAATACCCGATGATGGAAGTAGCTCTTATGGCTGGAGATTGCAGGTTTATGACTGATCCTATTTGGGAAGAACAATGATGGCAGGCATTCGAGCTTTATTTATGTACTTGTGGCTGCAGCTGGACTGGGTGAGCAGAGGAGAGAGTGTGGGGCTGCATCTTCCTACCCTGAGTGTCCAGGAGGGTGACAACTCTATTATCAACTGTGCTTATTCAAACAGCGCCTCAGACTACTTCATTTGGTACAAGCAAGAATCTGGAAAAGGTCCTCAATTCATTATAGACATTCGTTCAAATATGGACAAAAGGCAAGGCCAAAGAGTCACCGTTTTATTGAATAAGACAGTGAAACATCTCTCTCTGCAAATTGCAGCTACTCAACCTGGAGACTCAGCTGTCTACTTTTGTGCAGAGGCATCCGAGGGGAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACGGTCATTCCAAATATCCAGAACCCTGACCCTGCCGTGTACCAGCTGAGAGACTCTAAATCCAGTGACAAGTCTGTCTGCCTATTCACCGATTTTGATTCTCAAACAAATGTGTCACAAAGTAAGGATTCTGATGTGTATATCACAGACAAAACTGTGCTAGACATGAGGTCTATGGACTTCAAGAGCAACAGTGCTGTGGCCTGGAGCAACAAATCTGACTTTGCATGTGCAAACGCCTTC"</span>

<span class="co"># at indices 3 and 4 are plots (and the respective data frames) for V(D)JC or V(D)J alignments</span>
<span class="co"># these plots are made with a custom function igsc::MultiplePairwiseAlignmentsToOneSubject()</span>
<span class="co"># names on the y-axis show the pattern name its the full length in nt</span>
<span class="co"># number-annotations in the plot indicate the first and last nt of the alignment of the respective pattern</span>
<span class="co"># vertical dashed lines indicate the lower and upper limit of all patterns aligned to the subject</span>
<span class="va">align_list_TRA</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<p><img src="prepare_tcr_seq_data_files/figure-html/unnamed-chunk-13-2.png" width="768"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># the last sub-index contains the lower and upper limit as integers</span>
<span class="va">align_list_TRA</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; [1] 139 771</span>

<span class="co"># the last index of align_list_TRA contains the V(D)J-sequence of your TCR (sequence between the vertical lines of align_list_TRA[[4]][[2]])</span>
<span class="co"># if the alignment plots are as expected this sequence may be used for further processing</span>
<span class="va">align_list_TRA</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="co">#&gt; [1] "ATGGCAGGCATTCGAGCTTTATTTATGTACTTGTGGCTGCAGCTGGACTGGGTGAGCAGAGGAGAGAGTGTGGGGCTGCATCTTCCTACCCTGAGTGTCCAGGAGGGTGACAACTCTATTATCAACTGTGCTTATTCAAACAGCGCCTCAGACTACTTCATTTGGTACAAGCAAGAATCTGGAAAAGGTCCTCAATTCATTATAGACATTCGTTCAAATATGGACAAAAGGCAAGGCCAAAGAGTCACCGTTTTATTGAATAAGACAGTGAAACATCTCTCTCTGCAAATTGCAGCTACTCAACCTGGAGACTCAGCTGTCTACTTTTGTGCAGAGGCATCCGAGGGGAACACCGGTAACCAGTTCTATTTTGGGACAGGGACAAGTTTGACGGTCATTCCAA"</span></code></pre></div>
</div>
<div id="concanetate-a-tcr-cassette" class="section level2">
<h2 class="hasAnchor">
<a href="#concanetate-a-tcr-cassette" class="anchor"></a>Concanetate a TCR cassette</h2>
<p>The obtained V(D)J-sequences can then be used to concatenate the complete TCR (attach the constant gene segment) or a whole expression cassette with additional elements like a P2A element, restriction sites, etc. Below one possible cassette is created. A fasta file with cassette elements is included in this package.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcr_cassette_elements</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">:::</span><span class="fu">read_fasta</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"TCR_cassette_elements.fasta"</span>, package <span class="op">=</span> <span class="st">"igsc"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># concatenate</span>
<span class="va">concat_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html">toupper</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"NotI_site"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># left restriction site</span>
                             <span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"Kozak_sequence_noATG"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># eukaryotic translation initiation</span>
                             <span class="va">align_list_TRB</span><span class="op">[[</span><span class="st">"V.to.J.TCR.seq"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># TRB V-&gt;J chain</span>
                             <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"mTRBC2_Ser57Cys_STOPcodon_removed"</span><span class="op">]</span><span class="op">]</span>,
                                    <span class="fl">2</span>,
                                    <span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span><span class="op">(</span><span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"mTRBC2_Ser57Cys_STOPcodon_removed"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, <span class="co"># mTRBC, see info below</span>
                             <span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"gsg_linker"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># linker that provides flexibility for more efficient cleavage</span>
                             <span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"P2A_element"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># a cleavable linker</span>
                             <span class="va">align_list_TRA</span><span class="op">[[</span><span class="st">"V.to.J.TCR.seq"</span><span class="op">]</span><span class="op">]</span>, <span class="co"># TRA V-&gt;J chain</span>
                             <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"mTRAC_Thr48Cys"</span><span class="op">]</span><span class="op">]</span>,
                                    <span class="fl">2</span>,
                                    <span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span><span class="op">(</span><span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"mTRAC_Thr48Cys"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, <span class="co">#mTRAC, see info below</span>
                             <span class="va">tcr_cassette_elements</span><span class="op">[[</span><span class="st">"EcoRI_site"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># right restriction site</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">concat_seq</span><span class="op">)</span>
<span class="co">#&gt; [1] "GCGGCCGCGCCACCAGGATCTGAGAAATGTGACTCCACCCAAGGTCTCCTTGTTTGAGCCATCAAAAGCAGAGATTGCAAACAAACAAAAGGCTACCCTCGTGTGCTTGGCCAGGGGCTTCTTCCCTGACCACGTGGAGCTGAGCTGGTGGGTGAATGGCAAGGAGGTCCACAGTGGGGTCTGCACGGACCCTCAGGCCTACAAGGAGAGCAATTATAGCTACTGCCTGAGCAGCCGCCTGAGGGTCTCTGCTACCTTCTGGCACAATCCTCGAAACCACTTCCGCTGCCAAGTGCAGTTCCATGGGCTTTCAGAGGAGGACAAGTGGCCAGAGGGCTCACCCAAACCTGTCACACAGAACATCAGTGCAGAGGCCTGGGGCCGAGCAGACTGTGGAATCACTTCAGCATCCTATCATCAGGGGGTTCTGTCTGCAACCATCCTCTATGAGATCCTACTGGGGAAGGCCACCCTATATGCTGTGCTGGTCAGTGGCCTGGTGCTGATGGCCATGGTCAAGAAAAAAAATTCCGGCAGCGGCGCCACCAACTTCAGCCTGCTGAAGCAGGCCGGCGACGTGGAAGAGAACCCCGGGCCCACATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAATGCGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAGACAAGCTTCACCTGCCAAGATATCTTCAAAGAGACCAACGCCACCTACCCCAGTTCAGACGTTCCCTGTGATGCCACGTTGACTGAGAAAAGCTTTGAAACAGATATGAACCTAAACTTTCAAAACCTGTCAGTTATGGGACTCCGAATCCTCCTGCTGAAAGTAGCCGGATTTAACCTGCTCATGACGCTGAGGCTGTGGTCCAGTTGAGAATTC"</span></code></pre></div>
<p>Please note, that the the original sequences of constant chains have been altered. One codon each has been changed to a Cysteine (Ser57Cys, Thr48Cys) for TRB and TRA, respectively. This increases stability of the chain pairing. Moreover, in this example of human TCRs constant chain are taken from mouse which has several reasons. It decreases mispairing with (if present) endogenous TCR and positive transductants may easily be identified by a mAb for mTCRb (clone H57-597).</p>
</div>
<div id="find-matching-roughly-or-exact-cdr3s-in-the-vdjdb" class="section level2">
<h2 class="hasAnchor">
<a href="#find-matching-roughly-or-exact-cdr3s-in-the-vdjdb" class="anchor"></a>Find matching (roughly or exact) CDR3s in the <a href="https://vdjdb.cdr3.net">vdjdb</a>
</h2>
<p>Similar CDR3s are found by simple string matching with the levensthein algorithm. Additionally, differences between CDR3-‘peptides’ can be expressed more continuously by similarity based on the PAM30 substitution matrix, taking into account chemical properties of amino acids.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vdjdb</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"vdjdb.tsv.rds"</span>, package <span class="op">=</span> <span class="st">"igsc"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html">filter</a></span><span class="op">(</span><span class="va">Species</span> <span class="op">==</span> <span class="st">"HomoSapiens"</span><span class="op">)</span>

<span class="co"># filter for highly expanded clonotypes</span>
<span class="va">cl_long</span> <span class="op">&lt;-</span> <span class="va">cl_long</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html">filter</a></span><span class="op">(</span><span class="va">cl_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">cl_long</span><span class="op">$</span><span class="va">cl_name</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">cl_long</span><span class="op">$</span><span class="va">cl_name</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">12</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">matches</span> <span class="op">&lt;-</span> <span class="fu">igsc</span><span class="fu">::</span><span class="fu"><a href="../reference/vdjdb_hits.html">vdjdb_hits</a></span><span class="op">(</span>tcrs <span class="op">=</span> <span class="va">cl_long</span>,
                            vdjdb <span class="op">=</span> <span class="va">vdjdb</span>,
                            mapply_fun <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/parallel/mclapply.html">mcmapply</a></span>,
                            lapply_fun <span class="op">=</span> <span class="fu">parallel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>,
                            PAM30_similarity <span class="op">=</span> <span class="cn">F</span>,
                            mc.cores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="va">matches_append</span> <span class="op">&lt;-</span>
  <span class="va">cl_long</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">matches</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">vdjdb</span><span class="op">)</span>
<span class="co">#&gt; Joining, by = c("chain", "CDR3_aa_cr")</span>
<span class="co">#&gt; Joining, by = c("CDR3", "Gene")</span>


<span class="va">df</span> <span class="op">&lt;-</span>
  <span class="va">matches_append</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/distinct.html">distinct</a></span><span class="op">(</span><span class="va">CDR3</span>, <span class="va">CDR3_aa_cr</span>, <span class="va">lv</span>, <span class="va">Gene</span>, <span class="va">chain</span>, <span class="va">cl_name</span>, <span class="va">patient</span>, <span class="va">Epitope</span>, <span class="va">Epitope.gene</span>, <span class="va">Epitope.species</span>, <span class="va">Score</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidyr/man/drop_na.html">drop_na</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html">filter</a></span><span class="op">(</span><span class="va">lv</span> <span class="op">&lt;</span> <span class="fl">3</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/filter.html">filter</a></span><span class="op">(</span><span class="va">Score</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Score</span>, y <span class="op">=</span> <span class="va">lv</span>, fill <span class="op">=</span> <span class="va">Epitope.species</span>, color <span class="op">=</span> <span class="va">chain</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggbeeswarm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span>groupOnX <span class="op">=</span> <span class="cn">F</span>, cex <span class="op">=</span> <span class="fl">6</span>, size <span class="op">=</span> <span class="fl">3</span>, shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span>
  <span class="co">#geom_jitter(width = 0.15, height = 0.15, shape = 21, size = 3) +</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html">theme</a></span><span class="op">(</span>panel.grid.minor.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, panel.grid.minor.y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, strip.background <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>, text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html">element_text</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"Courier"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">fcexpr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/fcexpr/man/int_breaks.html">int_breaks</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>,<span class="fl">3.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/scale_continuous.html">scale_y_reverse</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">fcexpr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/fcexpr/man/int_breaks.html">int_breaks</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"grey60"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/guides.html">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/vars.html">vars</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">cl_name</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="prepare_tcr_seq_data_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># matches of high score (confidence level from vdjdb x-axis) and low levensthein distance (lv, y-axis) are most interesting</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Christopher Mark Skopnik.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
