<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compare CDR3s from sequencing to known CDR3s in the vdjdb — vdjdb_hits • igsc</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Compare CDR3s from sequencing to known CDR3s in the vdjdb — vdjdb_hits" />
<meta property="og:description" content="Matching CDR3s from sequencing with CDR3s of known antigen specificity may help to narrow potential antigen specificities
of one's sequenced TCRs. Export a tsv of all or selected entries here https://vdjdb.cdr3.net/search. Provide that tsv as
data frame (vdjdb). From vdjdb and tcrs distinct rows of tr_cols and cdr3_col are compared by levensthein distance
and if selected (PAM30_similarity) by amino acid similarity. Returned data frame may be used to join original columns
from vdjdb and tcrs followed by plotting with ggplot2 (see example). Original column names (vdj_tr_col, tcr_tr_col and
vdj_cdr3_col, tcr_cdr3_col) are kept in the returned data frame to allow easy subsequent joining of vdjdb and tcrs. It requires
though that these columns have different names in vdjdb and tcrs." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">igsc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9700</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/prepare_tcr_seq_data.html">prepare_tcr_seq_data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compare CDR3s from sequencing to known CDR3s in the vdjdb</h1>
    
    <div class="hidden name"><code>vdjdb_hits.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Matching CDR3s from sequencing with CDR3s of known antigen specificity may help to narrow potential antigen specificities
of one's sequenced TCRs. Export a tsv of all or selected entries here https://vdjdb.cdr3.net/search. Provide that tsv as
data frame (vdjdb). From vdjdb and tcrs distinct rows of tr_cols and cdr3_col are compared by levensthein distance
and if selected (PAM30_similarity) by amino acid similarity. Returned data frame may be used to join original columns
from vdjdb and tcrs followed by plotting with ggplot2 (see example). Original column names (vdj_tr_col, tcr_tr_col and
vdj_cdr3_col, tcr_cdr3_col) are kept in the returned data frame to allow easy subsequent joining of vdjdb and tcrs. It requires
though that these columns have different names in vdjdb and tcrs.</p>
    </div>

    <pre class="usage"><span class='fu'>vdjdb_hits</span><span class='op'>(</span>
  <span class='va'>vdjdb</span>,
  <span class='va'>tcrs</span>,
  vdj_tr_col <span class='op'>=</span> <span class='st'>"Gene"</span>,
  tcr_tr_col <span class='op'>=</span> <span class='st'>"chain"</span>,
  vdj_cdr3_col <span class='op'>=</span> <span class='st'>"CDR3"</span>,
  tcr_cdr3_col <span class='op'>=</span> <span class='st'>"CDR3_aa_cr"</span>,
  TRAxTRB <span class='op'>=</span> <span class='cn'>F</span>,
  max_lvdist <span class='op'>=</span> <span class='fl'>3</span>,
  PAM30_similarity <span class='op'>=</span> <span class='cn'>F</span>,
  mapply_fun <span class='op'>=</span> <span class='va'>mapply</span>,
  lapply_fun <span class='op'>=</span> <span class='va'>lapply</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>vdjdb</th>
      <td><p>data frame of TCR data from vdjdb; vdj_cdr3_col and vdj_tr_col must be present;
if not provided system.file("extdata", "vdjdb.tsv.rds", package = "igsc") is used (table downloaded 06/12/2021)</p></td>
    </tr>
    <tr>
      <th>tcrs</th>
      <td><p>data frame of TCR data from sequencing, tcr_cdr3_col and tcr_tr_col must be present; e.g. c_long from clonotype prep</p></td>
    </tr>
    <tr>
      <th>vdj_tr_col</th>
      <td><p>column which codes the TCR chain in the vdjdb reference data frame (should contain TRA and/or TRB only)</p></td>
    </tr>
    <tr>
      <th>tcr_tr_col</th>
      <td><p>column which codes the TCR chain in the tcrs data frame (should contain TRA and/or TRB only)</p></td>
    </tr>
    <tr>
      <th>vdj_cdr3_col</th>
      <td><p>column which codes the CDR3s in the vdjdb reference data frame</p></td>
    </tr>
    <tr>
      <th>tcr_cdr3_col</th>
      <td><p>column which codes the CDR3s in the tcrs data frame</p></td>
    </tr>
    <tr>
      <th>TRAxTRB</th>
      <td><p>logical, apart from comparing TRA vs TRA and TRB vs TRB between tcrs and vdjdb also check for TRA vs TRB?</p></td>
    </tr>
    <tr>
      <th>max_lvdist</th>
      <td><p>maximum levensthein distance between CDR3s for return data frame (also filtering before PAM30 similarity calculation)</p></td>
    </tr>
    <tr>
      <th>PAM30_similarity</th>
      <td><p>logical whether to calculate the similarity (pairwise alignment) between CDR3s based on PAM30 substitution matrix</p></td>
    </tr>
    <tr>
      <th>mapply_fun</th>
      <td><p>mapply function for PAM30_similarity, suggested are mapply, pbapply::pbmapply, parallel::mcmapply</p></td>
    </tr>
    <tr>
      <th>lapply_fun</th>
      <td><p>lapply function for levensthein distance calculation, suggested are lapply, pblapply::pblapply, parallel::mclapply</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>arguments passed to mapply_fun and lapply_fun, most relevant: mc.cores (e.g parallel::detectCores())</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data frame of matched CDR3s from tcrs and vdjdb, column lv indicates the levensthein
between CDR3 from vdjdb and tcrs</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'><span class='co'># tcrs data frame with sequencing data from TCRs has to be created (cl_long is appropriate)</span></span>
<span class='r-in'><span class='va'>matches</span> <span class='op'>&lt;-</span> <span class='fu'>vdjdb_hits</span><span class='op'>(</span><span class='va'>vdjdb</span>, <span class='va'>tcrs</span>, mapply_fun <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='va'><a href='https://rdrr.io/r/parallel/mclapply.html'>mcmapply</a></span>, mc.cores <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span><span class='op'>(</span><span class='op'>)</span>, nthread <span class='op'>=</span> <span class='fu'>parallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/r/parallel/detectCores.html'>detectCores</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>matches_append</span> <span class='op'>&lt;-</span> <span class='va'>matches</span> <span class='op'>%&gt;%</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/mutate-joins.html'>left_join</a></span><span class='op'>(</span><span class='va'>tcrs</span><span class='op'>)</span> <span class='op'>%&gt;%</span> <span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/mutate-joins.html'>left_join</a></span><span class='op'>(</span><span class='va'>vdjdb</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='va'>df</span> <span class='op'>&lt;-</span></span>
<span class='r-in'><span class='va'>df</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'><span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/distinct.html'>distinct</a></span><span class='op'>(</span><span class='va'>CDR3</span>, <span class='va'>CDR3_aa_cr</span>, <span class='va'>lv</span>, <span class='va'>Gene</span>, <span class='va'>chain</span>, <span class='va'>cl_name</span>, <span class='va'>patient</span>, <span class='va'>Epitope</span>, <span class='va'>Epitope.gene</span>, <span class='va'>Epitope.species</span>, <span class='va'>Score</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'><span class='fu'>tidyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/tidyr/man/drop_na.html'>drop_na</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'><span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>cl_name</span> <span class='op'><a href='https://rdrr.io/r/base/match.html'>%in%</a></span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Madalyn"</span>, <span class='st'>"Mikaylla"</span>, <span class='st'>"Morgan"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'><span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>lv</span> <span class='op'>&lt;</span> <span class='fl'>3</span><span class='op'>)</span> <span class='op'>%&gt;%</span></span>
<span class='r-in'><span class='fu'>dplyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/dplyr/man/filter.html'>filter</a></span><span class='op'>(</span><span class='va'>Score</span> <span class='op'>&gt;</span> <span class='fl'>0</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'><span class='co'># use jitter for many hits (when beeswarms become overlapping)</span></span>
<span class='r-in'><span class='fu'>ggplot</span><span class='op'>(</span><span class='va'>df</span>, <span class='fu'>aes</span><span class='op'>(</span>x <span class='op'>=</span> <span class='va'>Score</span>, y <span class='op'>=</span> <span class='va'>lv</span>, color <span class='op'>=</span> <span class='va'>Epitope.species</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>ggbeeswarm</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html'>geom_beeswarm</a></span><span class='op'>(</span>groupOnX <span class='op'>=</span> <span class='cn'>F</span>, cex <span class='op'>=</span> <span class='fl'>4</span>, size <span class='op'>=</span> <span class='fl'>2.5</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='co'>#geom_jitter(width = 0.15, height = 0.15, size = 3) +</span></span>
<span class='r-in'> <span class='fu'>theme_bw</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>theme</span><span class='op'>(</span>panel.grid.minor.x <span class='op'>=</span> <span class='fu'>element_blank</span><span class='op'>(</span><span class='op'>)</span>, panel.grid.minor.y <span class='op'>=</span> <span class='fu'>element_blank</span><span class='op'>(</span><span class='op'>)</span>, strip.background <span class='op'>=</span> <span class='fu'>element_rect</span><span class='op'>(</span>fill <span class='op'>=</span> <span class='st'>"white"</span><span class='op'>)</span>, text <span class='op'>=</span> <span class='fu'>element_text</span><span class='op'>(</span>family <span class='op'>=</span> <span class='st'>"Courier"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>scale_x_continuous</span><span class='op'>(</span>breaks <span class='op'>=</span> <span class='fu'>fcexpr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/fcexpr/man/int_breaks.html'>int_breaks</a></span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>, limits <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.8</span>,<span class='fl'>3.2</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>scale_y_reverse</span><span class='op'>(</span>breaks <span class='op'>=</span> <span class='fu'>fcexpr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/fcexpr/man/int_breaks.html'>int_breaks</a></span><span class='op'>(</span>n <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>guides</span><span class='op'>(</span>fill <span class='op'>=</span> <span class='fu'>guide_legend</span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span><span class='op'>)</span>, color <span class='op'>=</span> <span class='fu'>guide_legend</span><span class='op'>(</span>override.aes <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span class='r-in'> <span class='fu'>facet_grid</span><span class='op'>(</span>rows <span class='op'>=</span> <span class='fu'>vars</span><span class='op'>(</span><span class='va'>chain</span><span class='op'>)</span>, cols <span class='op'>=</span> <span class='fu'>vars</span><span class='op'>(</span><span class='va'>patient</span>, <span class='va'>cl_name</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='op'>}</span></span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Christopher Mark Skopnik.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


